enhancements/ocp-coreos-layering/ocp-coreos-layering.md

```diff
@@ -0,0 +1,730 @@
+
+We do not desire to recreate the internal registry piece-by-piece, so the functionality provided by any MCO-supplied registry would be absolutely minimal and limited only to what would be required to support the MCO's layering operations.
+
+Related Spikes:
+https://issues.redhat.com/browse/MCO-281
+https://issues.redhat.com/browse/MCO-286
+
+#### Image reproducibility
+
+For images we would build in "Phase 1" and "Phase 2", the desire would eventually be that:
+
+- As long as we have the same customer base image and the same MCO content, the output of our image build should always be the same and result in the same image
+
+This would allow for easy change detection, as we'd be able to figure out if an image changed just by observing its digest. It would also make storing and preserving the MCO-built image less important -- we could always rebuild it, the only cost would be time.
+
+Unfortunately, this is not currently a "solved problem" in the container image space, and we as the MCO don't have full control of the problem space to solve it. All it takes is for a single byte to be out of place -- something as simple as a timestamp -- and it's unclear whether we will be able to solve this in the near future.
+
+For now we're assuming we will not have 100% reproducible images and are planning accordingly.
+
+#### Future Upgrade flow
+1. The "base image" will be part of the release image, the same as today.
+1. The CVO will replace a ConfigMap in the MCO namespace with the OS payload reference, as it does today.
+1. The MCO will update an `imagestream` object (e.g. `openshift-machine-config-operator/rhel-coreos`) when this ConfigMap changes.
+1. The MCO builds a container image for each `MachineConfigPool`, using the new base image
+
+1. Each machineconfig pool will support a `custom-external-coreos` imagestream for pulling externally built images
+1.MCD continues to perform drains and reboots,, but for the most part does not change files in `/etc` or install packages per node
+1. The Machine Configuration Server (MCS) will only serve a "bootstrap" Ignition configuration (pull secret, network configuration) sufficient for the node to pull the target container image.
+
+For clusters without any custom MachineConfig at all, the MCO will deploy the result of the `mco-coreos` build.
+
+<!-- current drawing: https://docs.google.com/drawings/d/1cGly8mjYDezVjUQiUEWd63IO_S2ZrgSbv1cT5X1KJIk/edit -->
+
+#### Preserving `MachineConfig`
+
+We cannot just drop `MachineConfig` as an interface to node configuration. Hence, the MCO will be responsible for starting new builds on upgrades or when new machine config content is rendered.
+
+For most configuration, instead of having the MCD write files on each node, it will be added into the image build run on the cluster.
```

Comment:
Implicitly, the MCD would still be able to do both to support RHEL nodes, right?

Reply:
yes. that is explicitly mentioned in Non-Goals section